KEP-4602: Crashloopbackoff alpha docs PR (#48499)
kubernetes/website: content/en/docs/concepts/workloads/pods/pod-lifecycle.md

[Sidecar containers and Pod lifecycle](/docs/concepts/workloads/pods/sidecar-containers/#sidecar-containers-and-pod-lifecycle)
explains the behaviour of `init containers` when specify `restartpolicy` field on it.

### Configurable container restart delay

> {{< feature-state for_k8s_version="v1.32" state="alpha" >}}

Reviewer: ```suggestion
{{< feature-state feature_state_name="KubeletCrashLoopBackOffMax" >}}
```

> {{< feature-state feature_state_name="KubeletCrashLoopBackOffMax" >}}

Reviewer 2: it seems that alpha is not rendering properly

- preview: https://deploy-preview-48499--kubernetes-io-main-staging.netlify.app/docs/concepts/workloads/pods/pod-lifecycle/#configurable-container-restart-delay

Author: Ah I think it had the wrong key (`state` instead of `gate`). Looks good now at https://deploy-preview-48499--kubernetes-io-main-staging.netlify.app/docs/concepts/workloads/pods/pod-lifecycle/#configurable-container-restart-delay after the fix in 333e896

With the alpha feature gate `KubeletCrashLoopBackOffMax` enabled, you can

> reconfigure the maximum delay between container restarts from the default of

Reviewer: how about
```suggestion
reconfigure the maximum delay between container start retries from the default of
```

> 300s (5 minutes). This configuration is set per node using kubelet
> configuration. In your [kubelet
> configuration](https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/),

Reviewer: Wrap a line wherever makes sense, but not in the middle of a link.
Use relative path rather than an absolute URL.

```suggestion
300s (5 minutes). This configuration is set per node using kubelet configuration.
In your [kubelet configuration](/docs/tasks/administer-cluster/kubelet-config-file/),
```

> set the `CrashLoopBackOff.MaxContainerRestartPeriod` field between `"1s"` and

Reviewer: Please double check the spelling of the field names.
My guess is that the field names starts with lower case letters.
It is upper case because we are programming using GoLang.
What we document here is what the user sees and uses.

```suggestion
set the `CrashLoopBackOff.MaxContainerRestartPeriod` field between `"1s"` and
```

`"300s"`. Delays on that node will still start at 10s and increase exponentially
by 2x each restart, until being capped at your configured maximum. If the
`CrashLoopBackOff.MaxContainerRestartPeriod` you configure is less than the

> default initial value of 10s, the initial delay will instead be set to the

Reviewer: Where does the '10s' mean, and does it come from?

Author: It comes from the existing behavior, described in the section above -- I've added that context in 
265a248. Does this address this comment fully?

Reviewer: So ... the default behavior is that, IIUC, if a container fails, kubelet restarts it.
If the container fails again (the second time), kubelet restarts it again.
If the container fails again (the 3rd time), kubelet does NOT restart the container immediately.
It waits for a backoff period, which defaults to 10 seconds, before the next retry, AND
the kubelet put the container into CrashLoopBackoff state.
The kubelet uses a exponential backoff mechanism to increase the backoff period.

Am I understanding this correct?

Author: Yes that's correct. And to to put one finer point on it there is one more step after that:

The kubelet stops increasing the backoff period at some point, by default 300s. All restarts thereafter will be delayed by that value (as it will no longer exponentially increase).

The new feature being documented here, when enabled, allows changes to that 300s maximum -- to whatever is configured by the user, per node.
